### DISCUSSION TITLE
Error in InventoryHP for decay of stable nuclides

### DISCUSSION THREAD

**Discussion by buresjan** [radioactivedecay/radioactivedecay#92] [Q&A] [answered]
Hello!

First of all huge thanks for this great open-source project!

I was attempting to benchmark `radioactivedecay` with other codes. In the benchmark I was also interested in how the `radioactivedecay` package would behave when asked to decay stable nuclides (half-life = 0s). When using `InventoryHP` I encountered an error (in the final activity values) when I included stable nuclides in the initial inventory. Below I provide minimal working examples for different trivial cases I examined:

**CASE#1**

In:
```python
inventory_info = {
'Ba-133': 1000.0,
'Cs-133': 1000.0  # stable; in decay chain of Ba-133
}
# construct inventory and let it decay
inv0 = rd.InventoryHP(inventory_info, 'Bq')
inv1 = inv0.decay(331862400.0, 's')
# print out result
print(inv1)
```
Out:
```
InventoryHP activities(Bq): {'Ba-133': nan, 'Cs-133': nan}, decay dataset: icrp107_ame2020_nubase2020
```
Note:
The time of decay is roughly equal to the half-live of Ba-133, hence the acitivity of Ba-133 should be ~500 Bq. Note that Cs-133 is part of the decay chain of Ba-133.

**CASE#2**

In:
```python
inventory_info = {
'Ba-133': 1000.0
}
# construct inventory and let it decay
inv0 = rd.InventoryHP(inventory_info, 'Bq')
inv1 = inv0.decay(331862400.0, 's')
# print out result
print(inv1)
```
Out:
```
InventoryHP activities(Bq): {'Ba-133': 500.1215971592449, 'Cs-133': 0.0}, decay dataset: icrp107_ame2020_nubase2020
```
Note:
In this case we obtain the correct answer, so the problem in CASE#1 seems to be caused by the presence of the stable nuclide Cs-133 in the initial inventory.

**CASE#3**

In:
```python
inventory_info = {
'Ba-133': 1000.0,
'Os-188': 1000.0, # stable; not in decay chain of Ba-133
}
# construct inventory and let it decay
inv0 = rd.InventoryHP(inventory_info, 'Bq')
inv1 = inv0.decay(331862400.0, 's')
# print out result
print(inv1)
```
Out:
```
InventoryHP activities(Bq): {'Ba-133': nan, 'Cs-133': nan, 'Os-188': nan}, decay dataset: icrp107_ame2020_nubase2020
```
Note:
The problem with `nan` values remains even if we include a stable nuclide (Os-188) that is not present in the decay chain of Ba-133, thus the problem is likely to be related solely to the presence of a stable nuclide in the intial inventory.

### QUESTION
Is this behaviour for the decay of stable nuclides desired? **(I suppose not, because when i tried to use the normal class `Inventory` - without high precision - the `nan` values were not present in the final inventory)**. So it seems to me that the problem is tied to the `InventoryHP` class.

More general question - what is the desired outcome of the decay of stable nuclides when using `radioactivedecay` package? Shouldn't at least a warning with a readable message be raised when a user tries to do this?

Thank you so much for your help!

**Comment by alexmalins** [comment, marked as answer]
hi @buresjan 
The issue here is you are initializing inventories with nonphysical values: it makes no sense to say the radioactivity of stable nuclides Cs-133 & Os-188 is 1000 Bq like in the examples. The radioactivity (number of decays per second) of any stable nuclide is by definition _0_ Bq.

Because of this, as soon you create the Inventory in Cases 1 & 3, internally the code performs a divide by zero (decay constant of Cs-133 & Os-188 is 0 s^-1), which creates `nan` values:

```
>>> rd.InventoryHP({'Ba133': 1000.0, 'Cs-133': 1000.0, 'Os-188': 1000.0}, 'Bq')
InventoryHP activities (Bq): {'Ba-133': 1000.0, 'Cs-133': nan, 'Os-188': nan}, decay dataset: icrp107_ame2020_nubase2020
```

With SymPy (backend maths library behind InventoryHP), those nans then cause everything to be nan in the results. For standard precision Inventory objects with a SciPy maths backend, it doesn't cause everything to fail but I still recommend against using an unphysical input:

```
>>> rd.Inventory({'Ba133': 1000.0, 'Cs-133': 1000.0, 'Os-188': 1000.0}, 'Bq').decay(331862400.0, 's')
Inventory activities (Bq): {'Ba-133': 500.1215971592449, 'Cs-133': nan, 'Os-188': nan}, decay dataset: icrp107_ame2020_nubase2020
```


If you want to include the quantity of a stable nuclide when you create the inventory, you need to input using a different unit from radioactivity (Bq), e.g. mass ("g", "kg", "mg"...) or moles ("mol", "kmol", "mmol"...) instead. Then everything works correctly:

```
>>> rd.InventoryHP({'Ba133': 1000.0, 'Cs-133': 1000.0, 'Os-188': 1000.0}, 'mol').decay(331862400.0, 's').moles()
{'Ba-133': 500.1215971592449, 'Cs-133': 1499.8784028407551, 'Os-188': 1000.0}
```


I guess we could improve the code though by adding a Warning or throwing an Error to users if they try and create an inventory by supplying a radioactivity (Bq value) for a stable nuclide.

**Reply by buresjan** [reply]
Hi!

Thank you so much for such a swift reply! I completely understand the rationale behind the implementation and the physical meaning of the problem. My confusion arose mainly from the different behaviour of Inventory and InventoryHP - this has now been cleared up, so thank you.  

Raising an Error or a Warning could be a good idea to clearly notify the user that he is trying to do something unphysical (I knew that decaying a stable nuclide does not make any sense, but I was interested in how the package would deal with this edge case). But I fully understand that some tasks are of higher importance than others 😄 Anyway, thank you again!

**Reply by alexmalins** [reply]
hey @buresjan !

> Raising an Error or a Warning could be a good idea to clearly notify the user that he is trying to do something unphysical

Check out the new release (v0.4.18) made just now. It raises a ValueError if the user tries to create an inventory with an activity for a stable nuclide, so prohibits the issue like above.

v0.4.18 is on PyPI now and will hit conda-forge in a few hours

**Reply by buresjan** [reply]
Hey again, @alexmalins!

I updated the package using pip and everything now looks fine, an error is thrown correctly...thank you! 👍
